### PR TITLE
Move pybind/python api to cpython [part2]

### PR DESCRIPTION
**Pull request by marigoold**
本PR完成了：
- 重命名了 https://github.com/Oneflow-Inc/oneflow/pull/8258 中as_number的一系列函数和宏，防止其与method中的函数命名冲突。比如重命名+运算符对应的函数 `PyTensorObject_add` 为 `PyTensorObject_nb_add`，防止其与 `Tensor.add` 对应的 `PyTensorObject_add` 冲突。
- 搬运了比 https://github.com/Oneflow-Inc/oneflow/pull/8258 中稍微复杂的部分api

这部分API的搬运思路如下：
- 实现了工具函数 `concat_self` ，用来把Tensor self和args拼接起来，以便生成的函数接口调用
- 对于参数解析比较复杂的函数，用一个宏 `DIRECT_PASS_FUNC` 来创建函数，统一调用 `functional_api.yaml.pybind.h` 里面的接口
- 对于参数解析比较简单的函数，在Python C api层就完成参数解析，然后调用functional接口
- 对于 `Tensor.int/float` 等转dtype的函数，用宏 `DATATYPE_FUNC` 创建（虽然只有四个），不调用 `functional_api.yaml.pybind.h` 里面的接口

其他：
- 对于部分functional_api.yaml中参数名和PyTorch对应函数参数名不对齐的函数，改成了和PyTorch相同的参数名
- cast函数原来没有pin_memory参数，现在增加
  

这个部分迁移的api如下：
- [ ]  Tensor.ndim = property(_ndim)
- [ ]  Tensor.numpy = _numpy
- [x]  Tensor.size = _size
- [ ]  Tensor.backward = _backward
- [ ]  Tensor.__setitem__ = _setitem
- [ ]  Tensor.__str__ = _str
- [ ]  Tensor.__repr__ = _repr
- [ ]  Tensor.__bool__ = is_nonzero
- [x]  Tensor.__iadd__ = _iadd
- [ ]  Tensor.addmm = _addmm
- [ ]  Tensor.__format__ = _format
- [ ]  Tensor.__index__ = _index
- [ ]  Tensor.__float__ = _scalar_float
- [ ]  Tensor.__int__ = _scalar_int
- [ ]  Tensor.__array__ = _numpy
- [ ]  Tensor.uniform_ = _uniform
- [ ]  Tensor.trunc_normal_ = _trunc_normal_
- [ ]  Tensor.kaiming_uniform_ = _kaiming_uniform
- [ ]  Tensor.kaiming_normal_ = _kaiming_normal
- [ ]  Tensor.xavier_normal_ = _xavier_normal
- [ ]  Tensor.xavier_uniform_ = _xavier_uniform
- [ ]  Tensor.orthogonal_ = _orthogonal
- [ ]  Tensor.normal_ = _normal
- [ ]  Tensor.fill_ = _fill
- [ ]  Tensor.copy_ = _copy
- [ ]  Tensor._meta_repr = _meta_repr
- [ ]  Tensor.floor_divide = _floor_divide
- [x]  Tensor.argmax = _argmax
- [x]  Tensor.argmin = _argmin
- [ ]  Tensor.argsort = _argsort
- [ ]  Tensor.argwhere = _argwhere
- [x]  Tensor.amin = _amin
- [x]  Tensor.atan2 = _atan2
- [x]  [Tensor.gt](http://tensor.gt/) = _gt
- [x]  [Tensor.ge](http://tensor.ge/) = _ge
- [x]  Tensor.cast = _cast
- [x]  Tensor.diag = _diag
- [x]  Tensor.diagonal = _diagonal
- [ ]  Tensor.add = _add
- [ ]  Tensor.add_ = _add_inplace
- [x]  Tensor.addcmul = _addcmul
- [x]  Tensor.addcmul_ = _addcmul_
- [ ]  Tensor.div = _truediv
- [ ]  Tensor.div_ = _truediv_inplace
- [ ]  Tensor.mul = _mul
- [ ]  Tensor.mul_ = _mul_
- [ ]  Tensor.sub = _sub
- [x]  Tensor.sub_ = _sub_inplace
- [ ]  Tensor.clamp = _clamp
- [ ]  Tensor.clamp_ = _clamp_
- [ ]  Tensor.clip = _clip
- [ ]  Tensor.clip_ = _clip_
- [ ]  Tensor.cpu = _cpu
- [ ]  Tensor.cuda = _cuda
- [ ]  Tensor.expand = _expand
- [ ]  Tensor.expand_as = _expand_as
- [x]  Tensor.fmod = _fmod
- [ ]  Tensor.flatten = _flatten
- [ ]  Tensor.flip = _flip
- [ ]  Tensor.in_top_k = _in_top_k
- [ ]  Tensor.index_select = _index_select
- [ ]  Tensor.minimum = _minimum
- [ ]  Tensor.maximum = _maximum
- [ ]  Tensor.new_empty = _new_empty
- [ ]  Tensor.new_ones = _new_ones
- [ ]  Tensor.new_zeros = _new_zeros
- [ ]  Tensor.pow = _pow
- [ ]  Tensor.var = _var
- [ ]  Tensor.std = _std
- [ ]  Tensor.matmul = _matmul
- [ ]  Tensor.softplus = _softplus
- [ ]  Tensor.tril = _tril
- [ ]  Tensor.triu = _triu
- [ ]  Tensor.where = _where
- [ ]  Tensor.norm = _norm
- [ ]  Tensor.transpose = _transpose
- [ ]  Tensor.permute = _permute
- [ ]  Tensor.local_to_global = _local_to_global
- [ ]  Tensor.global_to_global = _global_to_global
- [ ]  Tensor.to_global = _to_global
- [ ]  Tensor.relu = _relu
- [ ]  Tensor.relu_ = _relu_inplace
- [ ]  Tensor.softmax = _softmax
- [ ]  Tensor.log_softmax = _log_softmax
- [x]  Tensor.logical_and = _and
- [x]  Tensor.logical_or = _or
- [x]  Tensor.logical_not = _not
- [x]  Tensor.logical_xor = _xor
- [ ]  Tensor.roll = _roll
- [x]  Tensor.bmm = _bmm
- [ ]  Tensor.chunk = _chunk
- [ ]  Tensor.repeat = _repeat
- [ ]  Tensor.tile = _tile
- [ ]  Tensor.split = _split
- [ ]  Tensor.unbind = _unbind
- [ ]  Tensor.squeeze = _squeeze
- [ ]  Tensor.swapaxes = _swapaxes
- [ ]  Tensor.amax = _amax
- [ ]  Tensor.swapdims = _swapdims
- [ ]  Tensor.unfold = _unfold
- [ ]  Tensor.narrow = _narrow
- [ ]  Tensor.unsqueeze = _unsqueeze
- [ ]  [Tensor.to](http://tensor.to/) = _to
- [ ]  Tensor.half = _half
- [ ]  Tensor.gather = _gather
- [ ]  Tensor.all = _all
- [ ]  Tensor.any = _any
- [ ]  Tensor.T = property(_T)
- [ ]  Tensor.masked_fill = _masked_fill
- [ ]  Tensor.masked_select = _masked_select
- [ ]  Tensor.eq = _eq
- [x]  [Tensor.ne](http://tensor.ne/) = _ne
- [ ]  Tensor.item = _item
- [x]  [Tensor.lt](http://tensor.lt/) = _lt
- [x]  Tensor.le = _le
- [ ]  Tensor.to_local = _to_local
- [ ]  Tensor.reshape = _reshape
- [ ]  Tensor.reshape_as = _reshape_as
- [ ]  Tensor.view = _view
- [ ]  Tensor.view_as = _view_as
- [ ]  Tensor.sort = _sort
- [ ]  Tensor.type_as = _type_as
- [ ]  Tensor.tolist = _tolist
- [x]  Tensor.int = _int
- [x]  Tensor.long = _long
- [x]  Tensor.float = _float
- [x]  Tensor.double = _double
- [ ]  Tensor.is_floating_point = _is_floating_point
- [ ]  Tensor.topk = _topk
- [ ]  Tensor.nms = _nms
- [ ]  Tensor.nonzero = _nonzero
- [ ]  Tensor.max = _max
- [ ]  Tensor.min = _min
- [ ]  Tensor.median = _median
- [ ]  Tensor.sum = _sum
- [ ]  Tensor.mean = _mean
- [ ]  Tensor.prod = _prod
- [ ]  Tensor.is_consistent = _is_consistent
- [ ]  Tensor.to_consistent = _to_consistent
- [ ]  Tensor.new_tensor = _new_tensor
- [ ]  Tensor.cumsum = _cumsum
- [ ]  Tensor.cumprod = _cumprod

### CODE REVIEW

File: oneflow/api/python/framework/tensor_functions.cpp [hunk 1]
```diff
@@ -50,38 +58,38 @@ using functional::PyObjectPtr;
...
+NB_BINARY_FUNC(PyTensorObject_nb_fmod, functional::fmod);
+NB_BINARY_FUNC(PyTensorObject_nb_div, functional::div);
+NB_BINARY_FUNC(PyTensorObject_nb_and, functional::logical_and);
+NB_BINARY_FUNC(PyTensorObject_nb_xor, functional::logical_xor);
+NB_BINARY_FUNC(PyTensorObject_nb_or, functional::logical_or);
+NB_BINARY_FUNC(PyTensorObject_nb_floor_div, functional::floor_divide);
+NB_BINARY_FUNC(PyTensorObject_nb_true_div, functional::div);
+NB_BINARY_FUNC(PyTensorObject_nb_matrix_multiply, functional::matmul);
+
+static PyObject* PyTensorObject_nb_pow(PyObject* a, PyObject* b, PyObject* unsed) {
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor_functions.cpp:78]: 这个也可以用NB_BINARY_FUNC来定义

File: oneflow/api/python/framework/tensor_functions.cpp [hunk 2]
```diff
@@ -50,38 +58,38 @@ using functional::PyObjectPtr;
...
   PyObjectPtr tuple(PyTuple_Pack(2, a, b));
-  auto* result = functional::pow(NULL, tuple.get(), NULL);
+  PyObject* result = functional::pow(NULL, tuple.get(), NULL);
   if (PyErr_Occurred()) { throw py::error_already_set(); }
   return result;
   END_HANDLE_ERRORS
 }
 
-static PyObject* PyTensorObject_invert(PyObject* self) {
+static PyObject* PyTensorObject_nb_invert(PyObject* self) {
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor_functions.cpp:87]: 这个用NB_UNARY_FUNC来定义，下面的报错信息是和PyTorch对齐的么，对齐的话可以挪到functional实现里去

File: oneflow/api/python/framework/tensor_functions.cpp [hunk 3]
```diff
@@ -237,13 +283,102 @@ static PyObject* PyTensorObject_element_size(PyObject* self, PyObject* unused) {
...
+  HANDLE_ERRORS
+  PyObject* idx = Py_None;
+  static const char* keywords[2] = {"idx", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "|O:size", const_cast<char**>(keywords), &idx)) {
+    return NULL;
+  }
+  auto shape = PyTensor_Unpack(self)->shape();
+  PyObject* shape_object = TensorSize_NewFromShape(*shape);
+  if (idx == NULL || idx == Py_None) return shape_object;
+  return shape_object->ob_type->tp_as_mapping->mp_subscript(shape_object, idx);
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor_functions.cpp:303]: 这里最好写成`return TensorSize_NewFromShape(Shape{shape.At(idx)})`

File: oneflow/api/python/framework/tensor_functions.cpp [hunk 4]
```diff
@@ -237,13 +283,102 @@ static PyObject* PyTensorObject_element_size(PyObject* self, PyObject* unused) {
...
+  static const char* keywords[2] = {"other", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O:sub_", const_cast<char**>(keywords), &other)) {
+    return NULL;
+  }
+  PyObjectPtr dict(PyDict_New());
+  CHECK_OR_THROW(PyDict_SetItemString(dict.get(), "inplace", Py_True) > -1);
+  PyObjectPtr concat_args(PyTuple_Pack(2, self, other));
+  PyObject* result = functional::sub(NULL, concat_args.get(), dict.get());
+  if (PyErr_Occurred()) { throw py::error_already_set(); }
+  return result;
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor_functions.cpp:378]: 这里解析完参数之后可以直接调用前面定义的那个PyTensorObject_nb_inplace_sub

File: oneflow/api/python/framework/tensor_functions.cpp [hunk 5]
```diff
@@ -237,13 +283,102 @@ static PyObject* PyTensorObject_element_size(PyObject* self, PyObject* unused) {
...
+}
+
+static PyObject* PyTensorObject_matmul(PyObject* self, PyObject* args, PyObject* kwargs) {
+  HANDLE_ERRORS
+  PyObject* other = NULL;
+  static const char* keywords[2] = {"other", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O:matmul", const_cast<char**>(keywords),
+                                   &other)) {
+    return NULL;
+  }
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor_functions.cpp:358]: 这里不需要处理参数，self concat args之后和kwargs一起直接给functional::matmul就行

Comment (marigoold) [reply oneflow/api/python/framework/tensor_functions.cpp:358]: 这里的考虑是Tensor.matmul参数和functional::matmul不一样，Tensor.matmul只有两个参数，self和other
https://github.com/Oneflow-Inc/oneflow/blob/fda6ab7cb407e8ffd1d324ee176df8c622005eb7/python/oneflow/framework/tensor.py#L630-L631

File: oneflow/api/python/framework/tensor_functions.cpp (hunk 1 above)
Comment (marigoold) [reply oneflow/api/python/framework/tensor_functions.cpp:78]: pow函数原型要求是三个输入参数，比NB_BINARY_FUNC多了一个。用PyCFunction强转会有问题吗？如果没问题这里就可以归进去

File: oneflow/api/python/framework/tensor_functions.cpp (hunk 4 above)
Comment (marigoold) [reply oneflow/api/python/framework/tensor_functions.cpp:378]: 好的

File: oneflow/api/python/framework/tensor_functions.cpp (hunk 3 above)
Comment (marigoold) [reply oneflow/api/python/framework/tensor_functions.cpp:303]: 好的，这么做是为了节省一层调用吗？因为我这么写是看在python端，shape[idx]的调用和tuple差不多，就直接调用了它继承的这些方法

File: oneflow/api/python/framework/tensor_functions.cpp (hunk 1 above)
Comment (hjchen2) [reply oneflow/api/python/framework/tensor_functions.cpp:78]: oo不行，它的定义是ternaryfunc，那就这样吧

File: oneflow/api/python/framework/tensor_functions.cpp (hunk 3 above)
Comment (hjchen2) [reply oneflow/api/python/framework/tensor_functions.cpp:303]: 是的，现在这种方式创建了实际上创建了两个PyObject，一个是shape，一个是shape[idx]，shape_object还🉐️处理它的生命期问题，没有直接返回TensorSize_NewFromShape方便

File: oneflow/api/python/framework/tensor_functions.cpp (hunk 5 above)
Comment (hjchen2) [reply oneflow/api/python/framework/tensor_functions.cpp:358]: 好，那就这样也行，不需要改

File: oneflow/api/python/framework/tensor_functions.cpp (hunk 3 above)
Comment (marigoold) [reply oneflow/api/python/framework/tensor_functions.cpp:303]: > 是的，现在这种方式创建了实际上创建了两个PyObject，一个是shape，一个是shape[idx]，shape_object还🉐️处理它的减引用问题，没有直接返回TensorSize_NewFromShape方便

这样的话，idx是负数的时候会报错

Comment (hjchen2) [reply oneflow/api/python/framework/tensor_functions.cpp:303]: 那就处理一下负数，if (idx < 0) idx += tensor.ndim()

Comment (marigoold) [reply oneflow/api/python/framework/tensor_functions.cpp:303]: `return PyLong_FromLong(shape->At(idx));` 改成这样可以吗？因为tensor.size(idx)返回的是int，不是shape

File: oneflow/api/python/framework/tensor_functions.cpp [hunk 6]
```diff
@@ -237,13 +283,104 @@ static PyObject* PyTensorObject_element_size(PyObject* self, PyObject* unused) {
...
+  auto shape = PyTensor_Unpack(self)->shape();
+  if (idx_obj == NULL || idx_obj == Py_None) return TensorSize_NewFromShape(*shape);
+  int32_t idx = PyLong_AsLong(idx_obj);
+  CHECK_OR_THROW(idx >= -shape->NumAxes() && idx < shape->NumAxes())
+      << Error::IndexError() << "Dimension out of range (expected to be in range of ["
+      << -shape->NumAxes() << ", " << shape->NumAxes() - 1 << "], but got " << idx << ")";
+  idx = idx < 0 ? idx + shape->NumAxes() : idx;
+  return PyLong_FromLong(shape->At(idx));
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor_functions.cpp:308]:
```suggestion
  return PyLong_FromLongLong(shape->At(idx));
```

File: oneflow/api/python/framework/tensor_functions.cpp [hunk 7]
```diff
@@ -237,13 +283,104 @@ static PyObject* PyTensorObject_element_size(PyObject* self, PyObject* unused) {
...
+  HANDLE_ERRORS
+  PyObject* idx_obj = Py_None;
+  static const char* keywords[2] = {"idx", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "|O:size", const_cast<char**>(keywords),
+                                   &idx_obj)) {
+    return NULL;
+  }
+  auto shape = PyTensor_Unpack(self)->shape();
+  if (idx_obj == NULL || idx_obj == Py_None) return TensorSize_NewFromShape(*shape);
+  int32_t idx = PyLong_AsLong(idx_obj);
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor_functions.cpp:303]:
```suggestion
  int64_t idx = PyLong_AsLongLong(idx_obj);
```

**Review by hjchen2 — APPROVED** [review]

File: oneflow/api/python/framework/tensor_functions.cpp [hunk 8]
```diff
@@ -118,49 +126,51 @@ PyObject* PyTensorObject_inplace_pow(PyObject* a, PyObject* b, PyObject* unsed)
...
-    NULL,                             // nb_inplace_floor_div
-    PyTensorObject_inplace_true_div,  // nb_inplace_true_div
-
-    NULL,                            // nb_index
-    PyTensorObject_matrix_multiply,  // nb_matrix_multiply
-    NULL,                            // not implemented yet nb_inplace_matrix_multiply
+    PyTensorObject_nb_add,       // nb_add
+    PyTensorObject_nb_sub,       // nb_subtract
+    PyTensorObject_nb_mul,       // nb_multiply
+    PyTensorObject_nb_fmod,      // nb_remainder
```
Comment (Flowingsun007) [inline oneflow/api/python/framework/tensor_functions.cpp:132]: 这些nb_的缩写表示什么意思？

File: oneflow/api/python/framework/tensor_functions.cpp [hunk 9]
```diff
@@ -237,13 +283,104 @@ static PyObject* PyTensorObject_element_size(PyObject* self, PyObject* unused) {
...
+  PyObject* idx_obj = Py_None;
+  static const char* keywords[2] = {"idx", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "|O:size", const_cast<char**>(keywords),
+                                   &idx_obj)) {
+    return NULL;
+  }
+  auto shape = PyTensor_Unpack(self)->shape();
+  if (idx_obj == NULL || idx_obj == Py_None) return TensorSize_NewFromShape(*shape);
+  int64_t idx = PyLong_AsLongLong(idx_obj);
+  CHECK_OR_THROW(idx >= -shape->NumAxes() && idx < shape->NumAxes())
```
Comment (Flowingsun007) [inline oneflow/api/python/framework/tensor_functions.cpp:304]: `shape->NumAxes()`用了很多次，建议提取出来避免重复调用

File: oneflow/api/python/framework/tensor_functions.cpp (hunk 8 above)
Comment (marigoold) [reply oneflow/api/python/framework/tensor_functions.cpp:132]: 我认为是number的意思，python的文档里面就广泛用了这些缩写 https://docs.python.org/3/c-api/typeobj.html#sub-slots

Comment (Flowingsun007) [reply oneflow/api/python/framework/tensor_functions.cpp:132]: 没事，看到了

**Review by Flowingsun007 — APPROVED** [review]

File: python/oneflow/framework/tensor.py [hunk 10]
```diff
@@ -1243,18 +1216,11 @@ def RegisterMethods():
...
-    Tensor.lt = _lt
-    Tensor.le = _le
     Tensor.to_local = _to_local
     Tensor.sort = _sort
     Tensor.type_as = _type_as
     Tensor.tolist = _tolist
-    Tensor.int = _int
-    Tensor.long = _long
-    Tensor.float = _float
-    Tensor.double = _double
```
Comment (hjchen2) [inline python/oneflow/framework/tensor.py:1257]: 同时也要把这里不用的接口也删一下，比如_double、_float这些

**Pull request merged**